webrtc/stun/stun_connection.nim

```diff
@@ -0,0 +1,203 @@
+  let localPreference = 65535'u32
+  let componentID = 1'u32
+  return (1 shl 24) * typePreference + (1 shl 8) * localPreference + (256 - componentID)
+
+proc getBindingRequest*(self: StunConn, username: string = ""): StunMessage =
+  ## Creates an encoded Binding Request
+  ##
+  result = StunMessage(msgType: StunBindingRequest)
+  self.rng[].generate(result.transactionId)
+
+  if username.len() == 0:
+    let ufrag = string.fromBytes(self.rng.genUfrag(32))
+    let p2pUsername = "libp2p+webrtc+v1/" & ufrag
+    result.attributes.add(UsernameAttribute.encode(p2pUsername & ":" & p2pUsername))
+  else:
+    result.attributes.add(UsernameAttribute.encode(username))
+
+  if self.iceControlling:
+    result.attributes.add(IceControlling.encode(self.iceTiebreaker))
+  else:
+    result.attributes.add(IceControlled.encode(self.iceTiebreaker))
+  result.attributes.add(Priority.encode(self.calculatePriority()))
+
+proc checkForError*(msg: StunMessage): Option[StunMessage] =
+  # Check for error from a BindingRequest message.
+  # Returns an option with some BindingErrorResponse if there is an error.
+  # Returns none otherwise.
+  # https://datatracker.ietf.org/doc/html/rfc5389#section-10.1.2
+  var res = StunMessage(msgType: StunBindingErrorResponse,
+                        transactionId: msg.transactionId)
+  if msg.getAttribute(AttrMessageIntegrity).isNone() or
+     msg.getAttribute(AttrUsername).isNone():
+    res.attributes.add(ErrorCode.encode(ECBadRequest))
+    return some(res)
+
+  # This check is related to the libp2p specification.
+  # Might be interesting to add a customizable function check.
+  let username = string.fromBytes(msg.getAttribute(AttrUsername).get().value)
+  let usersplit = username.split(":")
+  if usersplit.len() != 2 and not usersplit[0].startsWith("libp2p+webrtc+v1/"):
+    res.attributes.add(ErrorCode.encode(ECUnauthorized))
+    return some(res)
+
+  # https://datatracker.ietf.org/doc/html/rfc5389#section-15.9
+  var unknownAttr: seq[uint16]
+  for attr in msg.attributes:
+    let typ = attr.attributeType
+    if typ.isRequired() and typ notin StunAttributeEnum:
+      unknownAttr.add(typ)
+  if unknownAttr.len() > 0:
+    res.attributes.add(ErrorCode.encode(ECUnknownAttribute))
+    res.attributes.add(UnknownAttribute.encode(unknownAttr))
+    return some(res)
+
+  return none(StunMessage)
+
+# - Stun Messages Handler -
+
+proc stunMessageHandler(self: StunConn) {.async: (raises: [CancelledError]).} =
+  # Read indefinitely Stun messages from stunMsgs queue.
+  # Sends a BindingResponse or BindingResponseError after receiving a BindingRequest.
+  while true:
+    let message = await self.stunMsgs.popFirst()
+    try:
+      let decoded = StunMessage.decode(await self.stunMsgs.popFirst())
+      if decoded.msgType == StunBindingErrorResponse:
+        trace "Received a STUN error", decoded, remote = self.raddr
+        continue
+      elif decoded.msgType == StunBindingResponse:
+        # TODO: Handle it
```

Comment:
Why is this TODO here?

Reply:
Because, for the moment, we're not sending BindingRequest, so receiving BindingResponse is unlikely... And it would be very weird.
For the webrtc-direct, it may be necessary to implement BindingRequest sending. Some browsers may refuse connection because of this, but, from what I tested so far, it doesn't seem necessary.
For the webrtc, though, it is mandatory.

Reply:
I updated this TODO to make it more clear.